File: features/list/src/main/java/com/vp/list/viewmodel/ListViewModel.java
package com.vp.list.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import com.vp.list.model.ListItem;
import com.vp.list.model.SearchResponse;
import com.vp.list.service.SearchService;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import org.jetbrains.annotations.NotNull;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ListViewModel extends ViewModel {
    private final MutableLiveData<SearchResult> liveData = new MutableLiveData<>();
    private final SearchService searchService;

    private String currentTitle = "";
    private final List<ListItem> aggregatedItems = new ArrayList<>();

    @Inject
    ListViewModel(@NonNull SearchService searchService) {
        this.searchService = searchService;
    }

    public LiveData<SearchResult> observeMovies() {
        return liveData;
    }

    public void searchMoviesByTitle(@NonNull String title, int page) {
        if (page == 1 && !title.equals(currentTitle)) {
            aggregatedItems.clear();
            currentTitle = title;
            liveData.setValue(SearchResult.inProgress());
        } else {
            liveData.setValue(SearchResult.loadingMore(aggregatedItems));
        }
        executeSearchQuery(title, page);
    }

    private void executeSearchQuery(@NotNull String title, int page) {
        searchService.search(title, page).enqueue(new Callback<SearchResponse>() {
            @Override
            public void onResponse(
                @NonNull Call<SearchResponse> call,
                @NonNull Response<SearchResponse> response
            ) {

                try {
                    Thread.sleep(1000L);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                SearchResponse result = response.body();

                if (result != null) {
                    aggregatedItems.addAll(result.getSearch());
                    final SearchResult searchResult = SearchResult.success(
                        aggregatedItems,
                        result.getTotalResults()
                    );
                    liveData.setValue(searchResult);
                }
            }

            @Override
            public void onFailure(@NonNull Call<SearchResponse> call, @NonNull Throwable t) {
                liveData.setValue(SearchResult.error());
            }
        });
    }

    public void refreshMovies() {
        aggregatedItems.clear();
        liveData.setValue(SearchResult.inProgress());
        executeSearchQuery(currentTitle, 1);
    }
}

File: features/list/src/main/java/com/vp/list/viewmodel/ListViewModel.kt
package com.vp.list.viewmodel

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.vp.list.model.ListItem
import com.vp.list.model.SearchResponse
import com.vp.list.service.SearchService
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.ArrayList
import javax.inject.Inject

class ListViewModel @Inject internal constructor(
    private val searchService: SearchService
) : ViewModel() {
    private var currentTitle = ""
    private val aggregatedItems: MutableList<ListItem> = ArrayList()

    private val _movies = MutableLiveData<SearchResult>()
    val movies: LiveData<SearchResult> = _movies

    fun searchMoviesByTitle(title: String, page: Int) {
        if (page == 1 && title != currentTitle) {
            aggregatedItems.clear()
            currentTitle = title
            _movies.setValue(SearchResult.inProgress())
        } else {
            _movies.setValue(SearchResult.loadingMore(aggregatedItems))
        }
        executeSearchQuery(title, page)
    }

    private fun executeSearchQuery(title: String, page: Int) {
        searchService.search(title, page).enqueue(object : Callback<SearchResponse?> {
            override fun onResponse(
                call: Call<SearchResponse?>, response: Response<SearchResponse?>
            ) {
                val result = response.body()
                if (result != null) {
                    aggregatedItems.addAll(result.search)
                    val searchResult = SearchResult.success(aggregatedItems, result.totalResults)
                    _movies.value = searchResult
                }
            }

            override fun onFailure(call: Call<SearchResponse?>, t: Throwable) {
                _movies.value = SearchResult.error()
            }
        })
    }

    fun refreshMovies() {
        aggregatedItems.clear()
        _movies.value = SearchResult.inProgress()
        executeSearchQuery(currentTitle, 1)
    }
}
